node/src/actors/chain_manager/mining.rs

```diff
@@ -548,7 +551,16 @@ impl ChainManager {
                         })
                         .map(move |res|
                             res.map(move |result| match result {
-                                    Ok(value) => Ok((vrf_proof, collateral, value)),
+                                    Ok(value) => {
+                                if let RadonTypes::RadonError(error) = &value.result {
+                                    if error.inner() == &RadError::InconsistentSource {
+                                        log::warn!("Refraining not to commit to data request {} because the sources are apparently inconsistent", dr_pointer);
+                                        return Err(())
```

Comment:
I believe in this case it would be better for the network safety to commit the `RadError::InconsistentSource` than to refrain from commiting.

In https://github.com/witnet/witnet-rust/discussions/2050 we discuss possible attacks that arise from data requests that can only be solved by a small fraction of all the nodes. I believe the same arguments apply to the inconsistent source attack, because nodes that enable this inconsistent source attack protection will not participate in that requests.

Imagine an attacker that creates a data request that returns a bad value for some nodes in order to penalize them, the classical inconsistent source attack. If some nodes refrain from participating in that request, then the attacker will control a higher percentage of the quorum, leading to the attacker gaining more reputation. This is true until a majority of nodes enable inconsistent source attack protection, then the request will be resolved with a `RadError::InsufficientCommits`.

On the other hand, if the eligible nodes commit an error, then all the nodes will participate in solving this request, so it will be equally hard for the attacker to create an eligibility proof for this request than for any other request. If the majority of nodes have enabled inconsistent source attack protection, it is possible that the data request is solved with a radon error as a result, and in that case nobody gains reputation. And in the case where only a minority have enabled inconsistent source attack protection, there will be a few error commits in the data request, so the attack should still work, with a slightly lower profit.

Reply:
Hmmm... let me give that a thought. This is not a trivial point because what you are suggesting actually requires a protocol change to assign an identifier for `InconsistentSource` in `RadonErrors`.

Reply:
Of course I had already created such identifier, but this was a non-consensus-breaking change only because we are never putting that error into a commitment.

Reply:
True, in that case we could commit a `RadError::Unknown`, and in a future pull request add the logic to enable a new RadonError using TAPI.

Reply:
I have solved it like this:
```rust
.map_err(
    // The `InconsistentSource` error is mapped here for the sake of backwards
    // compatibility. Namely, to enable paranoid retrieval without having to
    // immediately introduce a breaking change that may jeopardize oracle
    // queries. The point of making the mapping here is to only affect actual
    // witnessing and committing, but not the `try_data_request` function, which
    // can rather use the `InconsistentSource` error for clarity.
    // TODO: pursue a WIP that introduces `InconsistentSource` as a proper
    //  RadonError at the protocol level.
    |err| match err {
        RadError::InconsistentSource => RadError::Unknown,
        other => other,
    },
)
```

Reply:
So it looks like the node already refrains from commiting errors in some cases, such as if the `RadManager` panics:

https://github.com/witnet/witnet-rust/blob/f6bcb721d11f6252daff6d72bdea6a385fb7d981/node/src/actors/chain_manager/mining.rs#L557-L561

And for some reason this pull request also enters that branch in my tests, so no commits are performed when the result is `RadError::InconsistentSource`. So not sure if we should change the behavior and commit an error in all cases, or only in case of inconsistent source, or don't commit even in case of inconsistent source.

Reply:
Interesting. As nodes are already refraining from committing in such cases... am I right to assume that we can change this part and start committing `RadonError::Unknown` in this "everything went wrong" case without breaking consensus?

In that case, we would start to see reveals with `RadonError::Unknown`, which currently are rare if not non-existing.

The thing is that the lack of commits caused by the code block you are pointing out is probably not even having an impact on oracle queries as of today, because those missing commits should be allowing other valid commits to take their place, under the assumption that there are always more eligible identities than necessary.

If that's the case, and we start to default to committing an error, that would most times lead to worse tallies than not doing so, because miners can't tell errors from values while at the commitment stage.

Reply:
> I right to assume that we can change this part and start committing RadonError::Unknown in this "everything went wrong" case without breaking consensus?

Should be safe, because nodes can always commit an error. So the result would be the same as a sudden increase in http errors or timeout errors.

> If that's the case, and we start to default to committing an error, that would most times lead to worse tallies than not doing so, because miners can't tell errors from values while at the commitment stage.

True, then I guess the right approach would be to keep the old behavior for now, and investigate about the consequences of changing it in the future. Would the ability to distinguish values from errors in the commit stage be useful? Or would it introduce even more attack vectors?

Reply:
> Would the ability to distinguish values from errors in the commit stage be useful? Or would it introduce even more attack vectors?

Idk, leaking that info at the commitment stage makes it to easy in theory for lazy witnesses not to perform the retrieval and simply copy the error from other commitments once they have seen `rf * consensus_requirement` of those (because at that point they will know in advance that the only value they can commit without being slashed must be an error).

Reply:
In the other hand, that would allow to skip the reveal phase :thinking:

Reply:
I was thinking about adding a flag to the commit like `is_error: bool`, this way lazy witnesses can only guess the error. And I believe if your error is different from the consensus error you miss the reward, or something like that, no sure because that logic is complex. That would allow miners to choose non-error commits, but it will also allow miners to choose error commits to sabotage a data request.